lib/saluki-components/src/destinations/datadog_metrics/endpoint/endpoints.rs

```diff
@@ -0,0 +1,114 @@
+use std::collections::HashMap;
+
+use regex::Regex;
+use serde::{Deserialize, Deserializer};
+use snafu::{ResultExt, Snafu};
+use url::Url;
+
+#[derive(Clone, Debug, Default)]
+pub struct AdditionalEndpoints {
+    pub endpoints: HashMap<String, Vec<String>>,
+}
+
+#[derive(Debug, Snafu)]
+#[snafu(context(suffix(false)))]
+pub enum EndpointError {
+    Parse { source: url::ParseError, domain: String },
+}
+
+#[derive(Default)]
+pub struct SingleDomainResolver {
+    domain: String,
+    api_keys: Vec<String>,
+}
+
+impl SingleDomainResolver {
+    pub fn domain(&self) -> &String {
+        &self.domain
+    }
+
+    pub fn api_keys(&self) -> &Vec<String> {
+        &self.api_keys
+    }
+
+    fn add_api_key(&mut self, api_key: String) {
+        self.api_keys.push(api_key);
+    }
+
+    fn set_domain(&mut self, domain: String) {
+        self.domain = domain;
+    }
+}
+
+pub fn create_single_domain_resolvers(
+    endpoints: &AdditionalEndpoints,
+) -> Result<Vec<SingleDomainResolver>, EndpointError> {
+    let mut resolvers = Vec::new();
+    for (domain, api_keys) in &endpoints.endpoints {
+        let new_domain = add_adp_version_to_domain(domain.to_string(), "adp".to_string())?;
+        let mut seen = HashMap::new();
+        let mut resolver = SingleDomainResolver::default();
+        for api_key in api_keys {
+            let trimmed_api_key = api_key.trim();
+            if trimmed_api_key.is_empty() || seen.contains_key(trimmed_api_key) {
+                continue;
+            }
+            seen.insert(trimmed_api_key, true);
+            resolver.add_api_key(trimmed_api_key.to_string());
+        }
+        if !resolver.api_keys.is_empty() {
+            resolver.set_domain(new_domain.trim_end_matches("/").to_string());
+            resolvers.push(resolver);
+        }
+    }
+    Ok(resolvers)
+}
+
+impl AdditionalEndpoints {
+    fn from_string(s: &str) -> Self {
+        let endpoints: HashMap<String, Vec<String>> =
+            serde_json::from_str(s).expect("Failed to parse string as JSON HashMap");
+        AdditionalEndpoints { endpoints }
+    }
+}
+
+impl<'de> Deserialize<'de> for AdditionalEndpoints {
+    fn deserialize<D>(deserializer: D) -> Result<AdditionalEndpoints, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        let s: String = Deserialize::deserialize(deserializer)?;
+        Ok(AdditionalEndpoints::from_string(&s))
+    }
+}
```

Comment:
With this implementation, we can _only_ support setting additional endpoints as a JSON-encoded string, so we wouldn't be able to use an existing `datadog.yaml` with `additional_endpoints` set.

I think what we should do here is:

- change `AdditionalEndpoints` to `AdditionalEndpoints(HashMap<String, Vec<String>>)`, and add `#[derive(Deserialize)]`, so that it can be natively deserialized from the existing structure that you would find in `datadog.yaml`
- add a `FromStr` impl for `AdditionalEndpoints` that calls into `serde_json::from_str`, to add the necessary support of parsing it from a JSON string
- finally, we use `serde-with` and the `PickFirst` helper to allow us to deserialize `AdditionalEndpoints` via either its native nested map form _or_ a JSON string: https://docs.rs/serde_with/latest/serde_with/struct.PickFirst.html#examples

Since we implemented `FromStr`, you could use their example with `DisplayFromStr` as-is as that's the codepath we'd want to utilize. We don't be about making this serializable, though.